delensalot/biases/iterbiasesN0N1.py

```diff
@@ -0,0 +1,424 @@
+
+        cls_ivfs_arr = utils.cls_dot([fal, dat_delcls, fal])
+        cls_ivfs = dict()
+        for i, a in enumerate(['t', 'e', 'b']):
+            for j, b in enumerate(['t', 'e', 'b'][i:]):
+                if np.any(cls_ivfs_arr[i, j + i]):
+                    cls_ivfs[a + b] = cls_ivfs_arr[i, j + i]
+
+        n_gg = nhl.get_nhl(qe_key, qe_key, cls_w, cls_ivfs, lmax_ivf, lmax_ivf, lmax_out=lmax_qlm)[0]
+        r_gg_true = qresp.get_response(qe_key, lmax_ivf, 'p', cls_w, cls_f, fal, lmax_qlm=lmax_qlm)[0]
+        N0_unbiased = n_gg * utils.cli(r_gg_true ** 2)  # N0 of QE estimator after rescaling by Rfid / Rtrue to make it unbiased
+
+        cls_plen_true['pp'] = cldd_true * utils.cli(np.arange(len(cldd_true)) ** 2 * np.arange(1, len(cldd_true) + 1, dtype=float) ** 2 / (2. * np.pi))
+        cls_plen_fid['pp'] = cldd_fid * utils.cli(np.arange(len(cldd_fid)) ** 2 * np.arange(1, len(cldd_fid) + 1, dtype=float) ** 2 / (2. * np.pi))
+        if 'wE' in version and it>0:
+            # Need to convert the template of the lensing power spectrum: Cldd*rho, into the reidual lensing of the map: Cldd*(1-rho)
+            cls_plen_true['pp'] =  cls_plen_true['pp'] *utils.cli( rho_sqd_phi) * (1. - rho_sqd_phi) 
+            cls_plen_fid['pp'] =  cls_plen_fid['pp'] *utils.cli( rho_sqd_phi) * (1. - rho_sqd_phi) 
+        elif 'wE' in version and it ==0:
+            cls_plen_true['pp'] =  cls2dls(cls_unl_true)[1] * utils.cli(np.arange(len(cldd_true)) ** 2 * np.arange(1, len(cldd_true) + 1, dtype=float) ** 2 /  (2. * np.pi))
+            cls_plen_fid['pp'] =  cls2dls(cls_unl_fid)[1] * utils.cli(np.arange(len(cldd_fid)) ** 2 * np.arange(1, len(cldd_fid) + 1, dtype=float) ** 2 /  (2. * np.pi))
+        
+        if 'wN1' in version:
+            if it == 0: print('Adding n1 in iterations')
+            from lensitbiases import n1_fft
+            from scipy.interpolate import UnivariateSpline as spl
+            lib = n1_fft.n1_fft(fal, cls_w, cls_f, np.copy(cls_plen_true['pp']), lminbox=50, lmaxbox=5000, k2l=None)
+            n1_Ls = np.arange(50, lmax_qlm + 1, 50)
+            if lmax_qlm not in n1_Ls:  n1_Ls = np.append(n1_Ls, lmax_qlm)
+            n1 = np.array([lib.get_n1(qe_key, L, do_n1mat=False) for L in n1_Ls])
+            N1_unbiased = spl(n1_Ls, n1_Ls ** 2 * (n1_Ls * 1. + 1) ** 2 * n1 / r_gg_true[n1_Ls] ** 2, k=2, s=0, ext='zeros')(np.arange(len(N0_unbiased)))
+            N1_unbiased *= utils.cli(np.arange(lmax_qlm + 1) ** 2 * np.arange(1, lmax_qlm + 2, dtype=float) ** 2)
+        else:
+            N1_unbiased = np.zeros(lmax_qlm + 1, dtype=float)
+
+        rho.append(rho_sqd_phi)
+        delcls_fid.append(cls_plen_fid)
+        delcls_true.append(cls_plen_true)
+
+    return delcls_fid, delcls_true
+
+
+
+def cls2N0N1(k, cls_cmb_filt, cls_cmb_dat, cls_noise_filt, cls_noise_dat, lmin_ivf, lmax_ivf, lmax_qlm, doN1mat=False):
```

Comment:
Some type annotation for this and all other functions would be great. Also docstring for some parameter missing, e.g. what is k? i think it is the same as qe_key in function above, can we uniformize the naming please?